statistics/handle/statscache.go

```diff
@@ -15,15 +15,274 @@ package handle
+	//SimpleStatsCacheType simple type
+	SimpleStatsCacheType
+)
+
+// DefStatsCacheType defines statistic cache type
+var DefStatsCacheType = RistrettoStatsCacheType
+
+// NewStatsCache returns a new statsCahce with capacity memoryLimit(initial 1G)
+func NewStatsCache(memoryLimit int64, tp statsCacheType) (StatsCache, error) {
+	switch tp {
+	case RistrettoStatsCacheType:
+		return newRistrettoStatsCache(memoryLimit)
+	case SimpleStatsCacheType:
+		return newSimpleStatsCache(memoryLimit), nil
+	}
+	return nil, errors.New("wrong statsCache type")
+}
+
+// newRistrettoStatsCache returns a new statsCahce with capacity memoryLimit(initial 1G)
+func newRistrettoStatsCache(memoryLimit int64) (*ristrettoStatsCache, error) {
+	// since newRistrettoStatsCache controls the memory usage by itself, set the capacity of
+	// the underlying LRUCache to max to close its memory control
+	sc := &ristrettoStatsCache{
+		memTracker: memory.NewTracker(memory.LabelForStatsCache, -1),
+	}
+	for i := range sc.tablesShards {
+		sc.tablesShards[i].data = make(map[int64]*statistics.Table)
+	}
+	sc.version = 0
+	cache, err := cache.NewCache(&cache.Config{
+		NumCounters: 1e7,         // number of keys to track frequency of (10M).
+		MaxCost:     memoryLimit, // maximum cost of cache (1GB).
+		BufferItems: 64,          // number of keys per Get buffer.
+		Cost: func(value interface{}) int64 {
+			return (value.(*statistics.Table).MemoryUsage())
+		},
+		OnEvict: func(key uint64, value interface{}) {
+			if t, ok := sc.Get(int64(key)); ok {
+				if value != nil && t.PhysicalID == int64(key) {
+					sc.memTracker.Consume(-t.MemUsage)
+					sc.Set(int64(key), t.CopyWithoutBucketsAndCMS())
+				}
+			}
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	sc.cache = cache
+	return sc, nil
+}
+
+// BytesConsumed returns the consumed memory usage value in bytes.
+// only used by test.
+func (sc *ristrettoStatsCache) BytesConsumed() int64 {
+	rndTbl := sc.GetAll()[0]
+	sc.Erase(rndTbl.PhysicalID)
+	// insert a table memory = 0 to make cache renew the memory
+	for i := 0; i < 10; i++ {
+		sc.Insert(rndTbl)
+		time.Sleep(10 * time.Millisecond)
+	}
+	return sc.memTracker.BytesConsumed()
+}
+
+// GetBytesLimit get the limits of memory.
+func (sc *ristrettoStatsCache) GetBytesLimit() int64 {
+	return sc.memTracker.GetBytesLimit()
+}
+
+// SetBytesLimit set new byteslimit
+func (sc *ristrettoStatsCache) SetBytesLimit(bytesLimit int64) {
+	sc.cache.SetNewMaxCost(bytesLimit)
+}
+
+// Close close the cache.
+func (sc *ristrettoStatsCache) Close() {
+	sc.cache.Clear()
+	sc.cache.Close()
+}
+
+// Clear clears the cache data.
+func (sc *ristrettoStatsCache) Clear() {
+	sc.cache.Clear()
+	sc.mu.Lock()
+	sc.version = 0
+	sc.mu.Unlock()
+	for i := range sc.tablesShards {
+		sc.tablesShards[i].Lock()
+		sc.tablesShards[i].data = make(map[int64]*statistics.Table)
+		sc.tablesShards[i].Unlock()
+	}
+	sc.memTracker = memory.NewTracker(memory.LabelForStatsCache, -1)
```

Comment:
sc.memTracker.Consume(sc.memTracker.BytesConsumed()), so that you can keep the ByteLimit for the next using